File: python/641. Design Circular Deque.py
class MyCircularDeque(object):

    def __init__(self, k):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        :type k: int
        """
        self.capacity = k+1
        self.que = [0]*self.capacity
        self.start = 0
        self.end = 0
        

    def insertFront(self, value):
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        self.que[self.start] = value
        newStart = (self.start - 1) % self.capacity
        if newStart < 0:
            newStart += self.capacity
        self.start = newStart

        return True 
        
        

    def insertLast(self, value):
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        self.que[self.end] = value
        newEnd = (self.end + 1) % self.capacity
        self.end = newEnd
        return True
        

    def deleteFront(self):
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        :rtype: bool
        """
        if self.isEmpty():
            return False
        self.start = (self.start + 1)%self.capacity
        return True
        

    def deleteLast(self):
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        :rtype: bool
        """
        if self.isEmpty():
            return False
        newEnd = (self.end - 1) % self.capacity     
        if newEnd < 0:
            newEnd += self.capacity
        self.end = newEnd
        return True
        

    def getFront(self):
        """
        Get the front item from the deque.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.que[self.start]        
        

    def getRear(self):
        """
        Get the last item from the deque.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        print self.que
        return self.que[self.end]
        

    def isEmpty(self):
        """
        Checks whether the circular deque is empty or not.
        :rtype: bool
        """
        return self.start==self.end
        

    def isFull(self):
        """
        Checks whether the circular deque is full or not.
        :rtype: bool
        """
        return (self.end + 1) % self.capacity==self.start
        


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: go/src/leetcode/641/641. Design Circular Deque.go
/**
@author ZhengHao Lou
Date    2022/08/15
*/
package main

type MyCircularDeque struct {
	front, end int
	k          int
	que        []int
}

func Constructor(k int) MyCircularDeque {
	return MyCircularDeque{
		que: make([]int, k+1),
		k:   k + 1,
	}
}

func (this *MyCircularDeque) InsertFront(value int) bool {
	if this.IsFull() {
		return false
	}
	this.front = (this.front - 1 + this.k) % this.k
	this.que[this.front] = value
	return true
}

func (this *MyCircularDeque) InsertLast(value int) bool {
	if this.IsFull() {
		return false
	}
	this.que[this.end] = value
	this.end = (this.end + 1) % this.k
	return true
}

func (this *MyCircularDeque) DeleteFront() bool {
	if this.IsEmpty() {
		return false
	}
	this.front = (this.front + 1) % this.k
	return true
}

func (this *MyCircularDeque) DeleteLast() bool {
	if this.IsEmpty() {
		return false
	}
	this.end = (this.end - 1 + this.k) % this.k
	return true
}

func (this *MyCircularDeque) GetFront() int {
	if this.IsEmpty() {
		return -1
	}
	return this.que[this.front]
}

func (this *MyCircularDeque) GetRear() int {
	if this.IsEmpty() {
		return -1
	}
	i := (this.end - 1 + this.k) % this.k
	return this.que[i]
}

func (this *MyCircularDeque) IsEmpty() bool {
	return this.front == this.end
}

func (this *MyCircularDeque) IsFull() bool {
	return (this.end+1)%this.k == this.front
}

func main() {
	que := Constructor(4)
	que.InsertLast(9)
	que.DeleteLast()
	que.GetRear()
	que.GetFront()
	
}
